homeassistant/components/mqtt/__init__.py

```diff
@@ -632,6 +629,11 @@ def _mqtt_on_disconnect(self, _mqttc, _userdata, result_code):
             time.sleep(wait_time)
             tries += 1
 
+    def _grouped_subscriptions(self):
+        """Return a dict with all subscriptions grouped by their topics."""
+        iterator = groupby(self.subscriptions, lambda sub: sub.topic)
```

Comment:
Make sure to sort `self.subscriptions` when using `groupby`

![image](https://user-images.githubusercontent.com/1444314/35945169-13fe0ac0-0c14-11e8-8178-4f5e59c08c6d.png)

I actually think that this function is quite overkill. For example, in `_mqtt_on_message`, it actually it more confusing what is going on compared to code that would just iterating over the list of subscriptions and matching each topic. I doubt it's also faster since you're now sorting the subscriptions, grouping them by topic and then converting them to a dictionary. Would that be faster than matching a few extra topics?

If you want to save on topic matching, you could consider converting `self.subscriptions` to be indexed by topic and as value have a list of subscriptions.

Reply:
> Make sure to sort self.subscriptions when using groupby

That's good to know. Thanks! (The tests probably didn't catch that because it doesn't really break anything - just makes things a bit slower)

> I doubt it's also faster since you're now sorting the subscriptions, grouping them by topic and then converting them to a dictionary. Would that be faster than matching a few extra topics?

I based this on my own usage of MQTT - and I use the availability topic *everywhere*, so I have lots of subscriptions to the same topic - so in that case this would be quicker since the regex matching is probably (much) slower than constructing a dictionary (in `O(N log N)`). 

(*If* we're really out for performance, I could also implement a trie - but I mean this is python after all 🙃)

> it actually it more confusing what is going on compared to code

That's definitely true.

> you could consider converting self.subscriptions to be indexed by topic

I actually had that in before, but the problem with that was that you'd then have the topic in the `Subscription` object **and** the dictionary keys - could potentially cause problems.

 > code that would just [iterate] over the list of subscriptions and [match] each topic

Yes, I believe this is the best option and I'll change the code accordingly.